DataProcessing/BybitFundingRateDownloader.cs

```diff
@@ -0,0 +1,333 @@
+    }
+
+
+    /// <summary>
+    /// Runs the instance of the object.
+    /// </summary>
+    /// <returns>True if process all downloads successfully</returns>
+    public bool Run()
+    {
+        var ratePerSymbol = new Dictionary<string, Dictionary<DateTime, decimal>>();
+        foreach (var date in GetProcessingDates())
+        {
+            foreach (var apiFundingRate in GetData(date))
+            {
+                var fundingTimestamp = long.Parse(apiFundingRate.FundingRateTimestamp, CultureInfo.InvariantCulture);
+                var fundingTime = Time.UnixMillisecondTimeStampToDateTime(fundingTimestamp);
+                if (!ratePerSymbol.TryGetValue(apiFundingRate.Symbol, out var dictionary))
+                {
+                    ratePerSymbol[apiFundingRate.Symbol] = dictionary = new();
+                }
+
+                var key = new DateTime(fundingTime.Year, fundingTime.Month, fundingTime.Day, fundingTime.Hour,
+                    fundingTime.Minute, fundingTime.Second);
+                dictionary[key] = decimal.Parse(apiFundingRate.FundingRate, CultureInfo.InvariantCulture);
+            }
+        }
+
+        foreach (var kvp in ratePerSymbol)
+        {
+            // USDC pairs have PERP in the name instead of USDC
+            SaveContentToFile(_destinationFolder, kvp.Key.Replace("PERP", "USDC"), kvp.Value);
+        }
+        
+        return true;
+    }
+
+    private IEnumerable<BybitFundingRate> GetData(DateTime date)
+    {
+        var start = (long)Time.DateTimeToUnixTimeStampMilliseconds(date.Date);
+        var end = (long)Time.DateTimeToUnixTimeStampMilliseconds(date.AddDays(1).Date);
+
+        var result = new List<BybitFundingRate>();
+
+        Parallel.ForEach(_perpetualFuturesExchangeInfos.Where(x => x.LaunchTimestamp <= end), exchangeInfo =>
+        {
+            
+            _indexGate.WaitToProceed();
+            var url =
+                $"{BybitApiEndpoint}/v5/market/funding/history?limit=200&symbol={exchangeInfo.Symbol}&startTime={start}&endTime={end}&category={exchangeInfo.Category}";
+            var data = url.DownloadData();
+
+            lock (result)
+            {
+                try
+                {
+                    result.AddRange(JsonConvert
+                        .DeserializeObject<ByBitResponse<BybitListResult<BybitFundingRate>>>(data, SerializerSettings)
+                        .Result.List);
+                }
+                catch (Exception)
+                {
+                    Log.Error($"GetData(): deserialization failed {data}");
+                    throw;
+                }
+            }
+        });
+        return result;
+    }
+
+
+    private IEnumerable<DateTime> GetProcessingDates()
+    {
+        if (_deploymentDate.HasValue)
+        {
+            return new[] { _deploymentDate.Value };
+        }
+        else
+        {
+            // everything
+            return Time.EachDay(new DateTime(2019, 9, 13), DateTime.UtcNow.Date);
```

Comment:
The start date should match when the bybit exchange started having this data available, trading the perpetual futures, tested for example `2019, 1, 1` and it seems the API is returning data too?

Reply:
The API is not returning data but actually returning with an error that there is no data available which is just being ignored. The question would also be there also how the downloader should handle hard errors. Just throw? Right now pretty much everything besides network errors is not being handled. 

I also double-checked when the first data is available, it's `2019,11,14`  will update it.